Dev customizable reports draw graph from data output

Anugraha-sutara opened the pull request:
<img src="https://github.com/user-attachments/assets/a278435f-cc83-46e8-80c3-845e862839b4" width="200" height="400">

Anugraha-sutara committed:
implement graph

mikedawson commented on lib-ui-compose/src/commonMain/kotlin/com/ustadmobile/libuicompose/view/report/graphs/Graphs.kt:
```diff
@@ -0,0 +1,330 @@
+    val (conversionFactor, unit) = remember(series) {
+        val maxY = series.flatMap { it.data.map { row -> row.yAxis } }.maxOrNull() ?: 0.0
+        when {
+            maxY >= 3_600_000 -> Pair(1.0 / 3_600_000, "hr")
```
No hardcoded strings! As per coding style: use localized strings.

mikedawson replied:
Still a problem

Anugraha-sutara committed:
change koalaplot version

mikedawson committed:
Merge branch 'dev-reports-mvvm' into dev-customizable-reports-draw-graph-from-data-output

mikedawson committed:
Add PlotlyKt for charts on web.

Anugraha-sutara committed:
Merge branch 'dev-reports-option-editing' into dev-customizable-reports-draw-graph-from-data-output

Anugraha-sutara committed:
update report edit screen as per wireframe

Anugraha-sutara committed:
update report edit screen

Anugraha-sutara committed:
change report icon

Anugraha-sutara committed:
update report edit screen

Anugraha-sutara committed:
add web graphs

Anugraha-sutara committed:
refactor

Anugraha-sutara committed:
add refactor

Anugraha-sutara committed:
resolve conflict

Anugraha-sutara committed:
update web report edit screen

Anugraha-sutara committed:
change report string

Anugraha-sutara committed:
refactor

Anugraha-sutara committed:
change report string

Anugraha-sutara committed:
change report string

Anugraha-sutara committed:
refactor

Anugraha-sutara committed:
refactor

Anugraha-sutara committed:
refactor

Anugraha-sutara committed:
refactor

Nikunjsharma0 commented on lib-ui-compose/src/commonMain/kotlin/com/ustadmobile/libuicompose/view/report/graphs/Graphs.kt:
```diff
@@ -0,0 +1,328 @@
+    // Collect all unique subgroups, treating null as a separate group
+    val allSubgroups = remember(series) {
+        series.flatMap { s ->
+            s.data.map { it.subgroup ?: "Default" }
```
please add in string file

Nikunjsharma0 commented on lib-ui-compose/src/commonMain/kotlin/com/ustadmobile/libuicompose/view/report/graphs/Graphs.kt:
```diff
@@ -0,0 +1,328 @@
+        allXValues.map { xValue ->
+            val subgroupValues = allSubgroups.map { subgroup ->
+                barSeries.flatMap { bs ->
+                    bs.data.filter { it.xAxis == xValue && (it.subgroup ?: "Default") == subgroup }
```
add in string file

Nikunjsharma0 commented on lib-ui-compose/src/commonMain/kotlin/com/ustadmobile/libuicompose/view/report/detail/ReportDetailScreen.kt:
```diff
@@ -0,0 +1,275 @@
+        CombinedGraph(
+            series = listOf(
+//                GraphSeries(SeriesType.BAR, barSeries1, "Bar Series 1"),
+//                GraphSeries(SeriesType.LINE, lineSeries, "Line Series 1"),
```
remove commented code

Nikunjsharma0 commented on lib-ui-compose/src/commonMain/kotlin/com/ustadmobile/libuicompose/view/report/detail/ReportDetailScreen.kt:
```diff
@@ -0,0 +1,275 @@
+            IconTextButton(
+                icon = Icons.Default.Share,
+                text = stringResource(MR.strings.share),
+                onClick = { /* Handle share */ }
```
share code missing

Nikunjsharma0 commented on lib-database/src/commonMain/kotlin/com/ustadmobile/lib/db/entities/ReportWithSeriesWithFilters.kt:
```diff
@@ -1,39 +1,39 @@
-    }
-}
\ No newline at end of file
+//package com.ustadmobile.lib.db.entities
```
remove this commented code

Nikunjsharma0 commented on lib-database/src/commonMain/kotlin/com/ustadmobile/core/db/dao/ReportDaoExt.kt:
```diff
@@ -1,9 +1,9 @@
-    val templateListToInsert = Report.FIXED_TEMPLATES.filter { it.reportUid !in uidsInserted }
-    replaceList(templateListToInsert)
-}
+//
```
remove this commented code

Nikunjsharma0 commented on core/src/commonMain/kotlin/com/ustadmobile/core/viewmodel/report/list/ReportListViewModel.kt:
```diff
@@ -0,0 +1,106 @@
+            ReportEditViewModel.DEST_NAME,
+            savedStateHandle[ARG_GO_TO_ON_REPORT_SELECTED]?.let {
+                mapOf(ARG_POPUP_TO_ON_REPORT_SELECTED to it)
+            } ?: emptyMap())
```
) it should end in  next line

Nikunjsharma0 commented on core/src/commonMain/kotlin/com/ustadmobile/core/util/ext/UmAppDatabaseExt.kt:
```diff
@@ -55,122 +54,122 @@ suspend fun <T: Person> UmAppDatabase.insertPersonAndGroup(
-//            MessageIdFormatter(
-//                    genderMap.mapKeys { it.key.toString() },
-//                    impl, context)
+//suspend fun UmAppDatabase.generateChartData(
```
remove if not in use

Anugraha-sutara committed:
Merge branch 'dev-reports-option-editing' into dev-customizable-reports-draw-graph-from-data-output

Anugraha-sutara committed:
resolve mr comments

poojaustad committed:
Graph page need to be updated

poojaustad committed:
Merge remote-tracking branch 'origin/dev-customizable-reports-draw-graph-from-data-output' into dev-customizable-reports-draw-graph-from-data-output

Anugraha-sutara committed:
refactor

Anugraha-sutara committed:
refactor

Anugraha-sutara committed:
fix cypress failure

Anugraha-sutara committed:
Merge branch 'dev-reports-option-editing' into dev-customizable-reports-draw-graph-from-data-output

# Conflicts:
#	core/src/commonMain/kotlin/com/ustadmobile/core/domain/report/model/ReportSeries2.kt

poojaustad committed:
Added screenshots for reports

Anugraha-sutara committed:
resolve conflict

Anugraha-sutara committed:
Merge remote-tracking branch 'origin/dev-customizable-reports-draw-graph-from-data-output' into dev-customizable-reports-draw-graph-from-data-output

Anugraha-sutara committed:
refactor

Anugraha-sutara committed:
update time range on report edit screen for android and desktop

Anugraha-sutara committed:
update time range on report edit screen for web

poojaustad committed:
Time range has been updated- maestro

poojaustad committed:
Time range has been updated- cypress

poojaustad committed:
Time range has been updated- maestro , test description

Anugraha-sutara committed:
add validation

Anugraha-sutara committed:
Merge remote-tracking branch 'origin/dev-customizable-reports-draw-graph-from-data-output' into dev-customizable-reports-draw-graph-from-data-output

Anugraha-sutara committed:
update web graph function

Anugraha-sutara committed:
resolve conflict

Anugraha-sutara committed:
Merge branch 'dev-reports-option-editing' into dev-customizable-reports-draw-graph-from-data-output

Anugraha-sutara committed:
resolve conflict

mikedawson commented on core/src/commonMain/kotlin/com/ustadmobile/core/viewmodel/report/detail/ReportDetailViewModel.kt:
```diff
@@ -50,48 +63,85 @@ class ReportDetailViewModel(
+                    launch {
+                        siteFlow.collect{
+                            val reportOptions = it?.let { optn ->
+                                optn.reportOptions?.takeIf { options -> options.isNotBlank() }
```
As per [coding style](https://github.com/UstadMobile/UstadMobile/blob/primary/CODING-STYLE.md#general-coding-style) ```Do not use null checks that fail silently and would lead to code that doesn't behave as expect with no logging output or exception```

This is a null check that would silently fail and mislead. If something didn't load here; it's going to just makeup default ReportOptions2? That will just show wrong data for no apparent reason. Please update to handle in accordance with coding style.

mikedawson replied:
Note; this is not the only please this coding style point applies - please check all your code / viewmodels and ensure that 
a) You are not silently or incorrectly handling null values
b) You are not ignoring / catching exceptions that should be thrown; if an exception is caught and then nothing is shown to the user and/or developer that is a problem.

mikedawson commented on core/src/commonMain/kotlin/com/ustadmobile/core/viewmodel/report/detail/ReportDetailViewModel.kt:
```diff
@@ -50,48 +63,85 @@ class ReportDetailViewModel(
+                            val reportOptions = it?.let { optn ->
+                                optn.reportOptions?.takeIf { options -> options.isNotBlank() }
+                                    ?.let { options ->
+                                        Json.decodeFromString(
```
Please use the json provided through the Dependency Injection (DI) - that is accessed using lower case ```json``` from a viewmodel

mikedawson commented on core/src/commonMain/kotlin/com/ustadmobile/core/viewmodel/report/detail/ReportDetailViewModel.kt:
```diff
@@ -50,48 +63,85 @@ class ReportDetailViewModel(
+                                    reportUid = reportUid,
+                                    reportOptions = reportOptions,
+                                    accountPersonUid = activeUserPersonUid,
+                                    timeZone = TimeZone.UTC
```
Use the system default timezone here.

mikedawson committed:
Report integration fixes:

1) ReportEditViewModel updated to set reportOwnerAccountPersonUid
2) ReportPeriod presets updated - last week, last 30 days, last 3 months were set to zero.

Anugraha-sutara committed:
resolve mr comments

mikedawson committed:
Merge branch 'dev-reports-mvvm' into dev-customizable-reports-draw-graph-from-data-output

# Conflicts:
#	core/src/commonMain/kotlin/com/ustadmobile/core/domain/report/model/ReportPeriod.kt
#	core/src/commonMain/kotlin/com/ustadmobile/core/domain/report/model/ReportTimeRange.kt
#	lib-test-common/src/commonMain/kotlin/com/ustadmobile/util/test/ext/umAppDatabaseSharedTestExt.kt

mikedawson committed:
Fix ReportEditViewModel / ReportPeriod compilation.

Anugraha-sutara committed:
resolve build failure

Anugraha-sutara committed:
update graph with real data

Anugraha-sutara committed:
refactor

Anugraha-sutara committed:
refactor

Anugraha-sutara committed:
refactor

Anugraha-sutara committed:
refactor

Anugraha-sutara committed:
refactor

Anugraha-sutara committed:
refactor

poojaustad committed:
maestro test update - commenting the assertion since empty data in the test

Anugraha-sutara committed:
add web side list screen

Anugraha-sutara committed:
resolve conflict

Anugraha-sutara committed:
add refactor

Anugraha-sutara committed:
add refactor

mikedawson commented on app-react/src/jsMain/kotlin/com/ustadmobile/view/report/detail/ReportDetailScreen.kt:
```diff
@@ -1,39 +1,231 @@
+            }
+
+            ListItem {
+                ListItemButton {
```
I think the share and export data aren't really implemented. We should remove buttons for anything not implemented.

Anugraha-sutara committed:
remove non-functional buttons

Anugraha-sutara committed:
add refactor

mikedawson commented on core/src/commonMain/kotlin/com/ustadmobile/core/viewmodel/report/detail/ReportDetailViewModel.kt:
```diff
@@ -45,53 +61,106 @@ class ReportDetailViewModel(
+                                val parsedOptions = when {
+                                    !optionsJson.isNullOrBlank() -> {
+                                        try {
+                                            Json.decodeFromString(
```
This should use json not Json so it uses the serializer from dependency injection

mikedawson commented on core/src/commonMain/kotlin/com/ustadmobile/core/viewmodel/report/detail/ReportDetailViewModel.kt:
```diff
@@ -45,53 +61,106 @@ class ReportDetailViewModel(
+                ) {
+                    val siteFlow = activeRepo.reportDao().findByUidLive(reportUid)
+                    launch {
+                        siteFlow.collect { report ->
```
Needs to use collectLatest here such that if a new value is emitted, the function should cancel
Should also use distinctBy (with last modified time) to prevent a situation where if the same value is emitted multiple times it would be interrupted.

e.g. should be distinctBy (to filter out duplicates that didn't really change) then collectLatest

mikedawson commented on core/src/commonMain/kotlin/com/ustadmobile/core/viewmodel/report/detail/ReportDetailViewModel.kt:
```diff
@@ -45,53 +61,106 @@ class ReportDetailViewModel(
+                    setLoadingState = true,
+                    permissionCheck = { true }
+                ) {
+                    val siteFlow = activeRepo.reportDao().findByUidLive(reportUid)
```
This is not a siteFlow - its reportFlow

mikedawson commented on core/src/commonMain/kotlin/com/ustadmobile/core/viewmodel/report/detail/ReportDetailViewModel.kt:
```diff
@@ -45,53 +61,106 @@ class ReportDetailViewModel(
+                                    accountPersonUid = activeUserPersonUid,
+                                    timeZone = TimeZone.currentSystemDefault()
+                                )
+                                runReport(request)
```
This needs to change to something like:


```
runReportUseCase(request).collect { reportResult -> 
                                    _uiState.update { prev ->
                                        prev.copy(reportResults = reportResult.results)
                                    }
                                }
```
Add try/catch - add somewhere in the UI to show the error message.

mikedawson commented on core/src/commonMain/kotlin/com/ustadmobile/core/viewmodel/report/edit/ReportEditViewModel.kt:
```diff
@@ -197,7 +203,13 @@ class ReportEditViewModel(
+                    newOptions.period,
+                    currentState.timeRangeError
+                ),
+                quantityError = quantityError
```
Update onfilterchanged:
```
private fun onFilterChanged(filter2: ReportFilter3, seriesId: Int) {
        if (filter2.reportFilterField != null) {
            _uiState.update { prev ->
                val updatedSeriesList = prev.reportOptions2.series.map { series ->
                    if (series.reportSeriesUid == seriesId) {
                        val existingFilters = series.reportSeriesFilters?.toMutableList() ?: emptyList()
                        val updatedFilters = existingFilters.replaceOrAppend(filter2) {
                            it.reportFilterUid == filter2.reportFilterUid
                        }
                        series.copy(reportSeriesFilters = updatedFilters)
                    } else {
                        series
                    }
                }

                prev.copy(
                    reportOptions2 = prev.reportOptions2.copy(
                        series = updatedSeriesList
                    )
                )
            }
            onEntityChanged(_uiState.value.reportOptions2)
        }
    }
    ```

mikedawson commented on core/src/commonMain/kotlin/com/ustadmobile/core/viewmodel/report/edit/ReportEditViewModel.kt:
YAxis and series errors need to be list/map type e.g. they are per series not one for the whole screen.

mikedawson replied:
onSeriesChanged: this was updating uiState twice - that needs fixed, and anywhere else that same mistake was made. change to something like
```
fun onSeriesChanged(updatedSeries: ReportSeries2) {
        onEntityChanged(
            _uiState.value.reportOptions2.let { reportOptions ->
                reportOptions.copy(
                    series = reportOptions.series.replace(updatedSeries) {
                        it.reportSeriesUid == updatedSeries.reportSeriesUid
                    }
                )
            }
        )
    }
```

mikedawson replied:
Calling onEntityChanged twice - has been fixed.
UI state error types are still singular, where they need to be a map or list

Anugraha-sutara committed:
add resolve comments

mikedawson commented on core/src/commonMain/kotlin/com/ustadmobile/core/viewmodel/report/edit/ReportEditViewModel.kt:
```diff
@@ -33,6 +35,8 @@ data class ReportEditUiState(
     val seriesTitleError: String? = null,
```
This is still an issue (as well as yAxisError): this is a singular field, when in fact, there are multiple series. This needs to be a map/list type where an error is linked to the uid of the series e.g. show the error only on the series that has an error.

mikedawson commented on lib-ui-compose/src/commonMain/kotlin/com/ustadmobile/libuicompose/view/report/detail/ReportDetailScreen.kt:
```diff
@@ -22,11 +48,147 @@ fun ReportDetailScreen(viewModel: ReportDetailViewModel) {
+            )
+        }
+    } else {
+        androidx.compose.material.Text("empty data") // need to change
```
1) we should use normal imports / not use the fully qualified package name in the function call 2) we must always use material3 imports.

mikedawson commented on lib-ui-compose/src/commonMain/kotlin/com/ustadmobile/libuicompose/view/report/detail/ReportDetailScreen.kt:
```diff
@@ -22,11 +48,147 @@ fun ReportDetailScreen(viewModel: ReportDetailViewModel) {
+
+@Composable
+fun BarGraphSampleScreen(
+    onShowDialog: () -> Unit = { },
```
onShowDialog no longer used.

mikedawson commented on core/src/commonMain/kotlin/com/ustadmobile/core/viewmodel/report/detail/ReportDetailViewModel.kt:
```diff
@@ -1,23 +1,40 @@
 
 data class ReportDetailUiState(
     val report: Report? = null,
+    val dialogVisible: Boolean = false,
```
Also needs removed.

mikedawson commented on core/src/commonMain/kotlin/com/ustadmobile/core/viewmodel/report/detail/ReportDetailViewModel.kt:
```diff
@@ -45,41 +63,72 @@ class ReportDetailViewModel(
+                            }
+                            .collectLatest { report ->
+                            try {
+                                val reportNonNull = report
```
Indentation needs fixed

mikedawson commented on core/src/commonMain/kotlin/com/ustadmobile/core/viewmodel/report/detail/ReportDetailViewModel.kt:
```diff
@@ -45,41 +63,72 @@ class ReportDetailViewModel(
+
+                            } catch (e: Exception) {
+                                _appUiState.update { it.copy(loadingState = NOT_LOADING) }
+                            }
```
Needs to catch exception AND show error message.

Anugraha-sutara committed:
add resolve comments

Anugraha-sutara committed:
Merge branch 'primary' into dev-reports-option-editing

Anugraha-sutara committed:
Merge branch 'dev-reports-option-editing' into dev-customizable-reports-draw-graph-from-data-output

Anugraha-sutara committed:
add domain use case that generates xAPI statements for a report

Anugraha-sutara committed:
add domain use case that generates xAPI statements for a report

Anugraha-sutara committed:
X-axis selection issue ,Y-axis selection logic

Anugraha-sutara committed:
fix Units not showing issue

Anugraha-sutara committed:
Implement functionality to edit existing filters

poojaustad committed:
Merge branch 'refs/heads/primary' into dev-customizable-reports-draw-graph-from-data-output

# Conflicts:
#	core/src/commonMain/resources/MR/base/strings.xml
#	test-end-to-end/webapp-cypress/cypress/support/commands.js

Anugraha-sutara committed:
resolve conflict

poojaustad committed:
Merge remote-tracking branch 'origin/dev-customizable-reports-draw-graph-from-data-output' into dev-customizable-reports-draw-graph-from-data-output

# Conflicts:
#	core/src/commonMain/resources/MR/base/strings.xml

Anugraha-sutara committed:
resolve conflict

poojaustad committed:
Merge remote-tracking branch 'origin/dev-customizable-reports-draw-graph-from-data-output' into dev-customizable-reports-draw-graph-from-data-output

Anugraha-sutara committed:
resolve conflict

Anugraha-sutara committed:
add username on testurl

poojaustad committed:
Http call for usage data added to cypress

Anugraha-sutara committed:
Fixed the screen titles as per the issue #1091

Anugraha-sutara committed:
Fixed issue #1092

poojaustad committed:
cypress test for reports test

Anugraha-sutara committed:
update edit screen as per issue  1093

Anugraha-sutara committed:
Merge branch 'dev-customizable-reports-draw-graph-from-data-output' of https://github.com/UstadMobile/UstadMobile into dev-customizable-reports-draw-graph-from-data-output

poojaustad committed:
maestro test - added usage data

Anugraha-sutara committed:
fix issue #1095

poojaustad committed:
E2e Tests cleanup with title validation

poojaustad committed:
Verify student user has no permission to see Reports tab

mikedawson commented:
@poojaustad @Anugraha-sutara The idea that a 'student user' should not see the reports tab is completely wrong. Users can be both e.g. a teacher can also have a student role in a training course.

Anyone can see the reports tab; their report data will only come from what they have permission to see.

poojaustad committed:
Updated test description for reports test

Anugraha-sutara committed:
resolve list screen ui issue for report

Anugraha-sutara committed:
resolve list screen ui issue for report

Anugraha-sutara committed:
resolve issue 1093

Anugraha-sutara committed:
update listscreen ui for report using grid

Anugraha-sutara committed:
update listscreen ui for report using grid

poojaustad committed:
cypress cleanup- added series

Anugraha-sutara committed:
update listscreen ui

Anugraha-sutara committed:
update listscreen ui

Anugraha-sutara committed:
update filter option

Anugraha-sutara committed:
update report list screen

Anugraha-sutara committed:
add refactor

poojaustad committed:
cypress and maestro report tests

poojaustad committed:
fix maestro report test

Anugraha-sutara committed:
Merge branch 'dev-reports-mvvm' into dev-customizable-reports-draw-graph-from-data-output

# Conflicts:
#	app-ktor-server/src/main/kotlin/com/ustadmobile/lib/rest/UmRestApplication.kt
#	core/src/commonMain/kotlin/com/ustadmobile/core/viewmodel/report/edit/ReportEditViewModel.kt
#	core/src/commonMain/kotlin/com/ustadmobile/core/viewmodel/report/list/ReportListViewModel.kt
#	core/src/commonMain/resources/MR/base/strings.xml
#	gradle/libs.versions.toml
#	lib-database/schemas/com.ustadmobile.core.db.UmAppDatabase/205.json
#	lib-ui-compose/src/commonMain/kotlin/com/ustadmobile/libuicompose/view/app/App.kt
#	test-end-to-end/webapp-cypress/cypress/support/commands.js

Anugraha-sutara committed:
resolve conflict

Anugraha-sutara committed:
Merge branch 'dev-reports-mvvm' into dev-customizable-reports-draw-graph-from-data-output

Anugraha-sutara committed:
fix test failure

Anugraha-sutara committed:
Merge branch 'dev-reports-mvvm' into dev-customizable-reports-draw-graph-from-data-output

poojaustad committed:
added maestro file clear_state_and_learning_space_existing_userflow.yaml on test 012

Anugraha-sutara committed:
refactor

mikedawson committed:
Update RunReportUseCaseDatabaseImpl to handle x axis labels for class.

Anugraha-sutara committed:
add dateformatter

Anugraha-sutara committed:
add refactor

Anugraha-sutara committed:
add refactor

Anugraha-sutara committed:
add refactor

Anugraha-sutara committed:
add refactor

mikedawson commented on lib-ui-compose/src/commonMain/kotlin/com/ustadmobile/libuicompose/view/report/graphs/Graphs.kt:
```diff
@@ -0,0 +1,454 @@
+import io.github.koalaplot.core.xygraph.FloatLinearAxisModel
+import io.github.koalaplot.core.xygraph.XYGraph
+
+private const val HOUR_UNIT = "hr"
```
These are hardcoded English strings. This is clearly not meeting coding styles and basic good practice.

mikedawson commented on lib-ui-compose/src/commonMain/kotlin/com/ustadmobile/libuicompose/view/report/detail/ReportDetailScreen.kt:
```diff
@@ -22,11 +52,214 @@ fun ReportDetailScreen(viewModel: ReportDetailViewModel) {
+@Composable
+fun getGenderLabel(rawValue: Any?): String {
+    return when (rawValue as? String) {
+        "2" -> stringResource(MR.strings.male)
```
Must use constants defined on the Person class - must not use a literal value here. As per coding style: https://github.com/UstadMobile/UstadMobile/blob/primary/CODING-STYLE.md#never-hardcode-any-literal-values

mikedawson commented on core/src/commonMain/kotlin/com/ustadmobile/core/util/report/ReportFormatter.kt:
```diff
@@ -0,0 +1,28 @@
+import kotlinx.datetime.toLocalDate
+
+
+object ReportFormatter {
```
Should be in existing package for reports: domain.report

mikedawson commented on lib-ui-compose/src/commonMain/kotlin/com/ustadmobile/libuicompose/view/report/graphs/Graphs.kt:
```diff
@@ -0,0 +1,454 @@
+    }
+
+    // Determine Y-axis unit and conversion factor
+    val maxY = remember(series) {
```
Shared logic should go in a logical place and be used on both web and compose

mikedawson commented on lib-ui-compose/src/commonMain/kotlin/com/ustadmobile/libuicompose/view/report/graphs/Graphs.kt:
```diff
@@ -0,0 +1,454 @@
+    }
+
+    // Collect all unique subgroups, treating null as a separate group
+    val allSubgroups = remember(series) {
```
Should also have a common implementation

mikedawson commented on lib-ui-compose/src/commonMain/kotlin/com/ustadmobile/libuicompose/view/report/graphs/Graphs.kt:
```diff
@@ -0,0 +1,454 @@
+        XYGraph(
+            modifier = modifier
+                .fillMaxSize(),
+            xAxisModel = FloatLinearAxisModel(
```
Using 0.5f hardcoded constant is (again) against the coding style. https://github.com/UstadMobile/UstadMobile/blob/primary/CODING-STYLE.md#never-hardcode-any-literal-values . If there is a specific reason why we should use that (e.g. from the graph library docs) then there should be a constant defined (as a const) and it should have a link to the doc / reason for its value in the comments.

mikedawson replied:
This should have a reference link

mikedawson commented on lib-ui-compose/src/commonMain/kotlin/com/ustadmobile/libuicompose/view/report/graphs/Graphs.kt:
```diff
@@ -0,0 +1,454 @@
+    val range = yRange.endInclusive - yRange.start
+    return when {
+        range == 0f -> 1f
+        yAxisLabel.equals(YAxisTypes.COUNT.name, ignoreCase = true) -> {
```
This would definitely break in any non-English language; and could break even in English if the user simply changed the label.

mikedawson commented on lib-ui-compose/src/commonMain/kotlin/com/ustadmobile/libuicompose/view/report/graphs/Graphs.kt:
```diff
@@ -0,0 +1,454 @@
+import androidx.compose.foundation.layout.size
+import androidx.compose.foundation.layout.width
+import androidx.compose.foundation.shape.RoundedCornerShape
+import androidx.compose.material.Text
```
This is importing from material(2) - this should be removed. Should import material3 and not use fully qualified names

mikedawson commented on lib-ui-compose/src/commonMain/kotlin/com/ustadmobile/libuicompose/view/report/graphs/Graphs.kt:
```diff
@@ -0,0 +1,454 @@
+                series.data.groupBy { it.subgroup }.forEach { (subgroup, dataPoints) ->
+                    val points = dataPoints.map {
+                        DefaultPoint(
+                            x = xValueToIndex[it.xAxis]!!.toFloat(),
```
Generally !! should never be used in production (non-test) code. If there is no good way to avoid using !!, then there must be a comment clearly explaining why the value could never really be null.

mikedawson commented on lib-ui-compose/src/commonMain/kotlin/com/ustadmobile/libuicompose/view/report/graphs/Graphs.kt:
```diff
@@ -0,0 +1,454 @@
+    ) {
+        Column(
+            modifier = Modifier.defaultItemPadding(
+                start = 5.dp,
```
Two problems: one shouldn't anyway be using hardcoded literals; secondly material ui is based on 8dp increments.

mikedawson commented on core/src/commonMain/kotlin/com/ustadmobile/core/viewmodel/report/list/ReportListViewModel.kt:
```diff
@@ -70,11 +87,40 @@ class ReportListViewModel(
+    fun runReport(report: Report): Flow<ReportDataResult> = flow {
+        try {
+            val reportOptions = report.reportOptions?.let {
+                Json.decodeFromString<ReportOptions2>(it)
```
This needs to use the json NOT Json (e.g. the one we have from dependency injection)

mikedawson commented on core/src/commonMain/kotlin/com/ustadmobile/core/viewmodel/report/list/ReportListViewModel.kt:
```diff
@@ -2,24 +2,39 @@ package com.ustadmobile.core.viewmodel.report.list
     val addSheetOrDialogVisible: Boolean = false,
 )
 
+data class ReportDataResult(
```
This class should not exist: we already have RunReportUseCase.RunReportResult which includes the request (including options) and the reportrows

Anugraha-sutara committed:
resolve MR comments related to graph fun

Anugraha-sutara committed:
add code refactor

Anugraha-sutara committed:
add code refactor

mikedawson commented on lib-ui-compose/src/commonMain/kotlin/com/ustadmobile/libuicompose/view/report/graphs/Graphs.kt:
```diff
@@ -0,0 +1,481 @@
+        legend = { if (!compactMode) CombinedLegend(series, colorMap, reportOptions) },
+        legendLocation = LegendLocation.BOTTOM
+    ) {
+        XYGraph(
```
Must add comment with link to docs/samples from library e.g. https://koalaplot.github.io/0.5/docs/xygraphs/bar_plots/

mikedawson commented on lib-ui-compose/src/commonMain/kotlin/com/ustadmobile/libuicompose/view/report/graphs/Graphs.kt:
```diff
@@ -0,0 +1,481 @@
+private const val MS_IN_HOUR = 3_600_000
+private const val MS_IN_MINUTE = 60_000
+private const val MS_IN_SECOND = 1_000
+
```
@Anugraha-sutara please create a new CombinedGraph function, starting with the Koala plot example code, only for the bar chart.

Please keep it as simple and as close to the example as possible.

Anugraha-sutara committed:
add new CombinedGraph fun

mikedawson committed:
Refactor

mikedawson committed:
Remove duplicate classes that should not exist: ReportResultQueryRow, GraphSeries

Update RunReportResult to include functions/structures required to get info commonly needed for graphs e.g. distinct axis.

Anugraha-sutara committed:
a) removed use of classes ReportResultQueryRow and GraphSeries
b) Created preview function

Anugraha-sutara committed:
a) removed use of classes ReportResultQueryRow and GraphSeries
b) Created preview function

Anugraha-sutara committed:
fix build failure

Anugraha-sutara committed:
add refactor

Anugraha-sutara committed:
Used RunReportResult as the argument type for the CombinedGraph function

mikedawson committed:
Additional refactoring for clarity of Graph UI (Compose).

Anugraha-sutara committed:
a) Add line charts back to compose
b) Refactor React/web version

mikedawson commented on lib-ui-compose/src/commonMain/kotlin/com/ustadmobile/libuicompose/view/report/graphs/CombinedGraph.kt:
```diff
@@ -0,0 +1,134 @@
+
+                val dataPoints = statementReportRow.map { point ->
+                    Point(point.xAxis, point.yAxis.toFloat())
+                }.sortedBy { it.x }
```
I think this logic is flawed: sorting by the xAxis could give a different order. Eg. when showing data by month, the x axis (formatted) might be "Jan 2025", "Feb 2025", which if sorted alphabetically, will yield nonsense.

Anugraha-sutara committed:
Refactor React/web version

Anugraha-sutara committed:
Refactor React/web version

mikedawson commented on lib-ui-compose/src/commonMain/kotlin/com/ustadmobile/libuicompose/view/report/graphs/CombinedGraph.kt:
```diff
@@ -0,0 +1,162 @@
+                }
+            }
+        }
+
```
@Anugraha - please add a reference to Koala plot documentation

mikedawson replied:
Still required

mikedawson commented on app-react/src/jsMain/kotlin/com/ustadmobile/view/report/graph/ReportGraph.kt:
```diff
@@ -0,0 +1,123 @@
+                            "staticPlot" put isCompact
+                        }
+                    }
+                ) {
```
Anugraha - please add a reference to official example or docs link in comment

mikedawson replied:
Still required

mikedawson committed:
Minor tidying.

mikedawson committed:
Add a more elegant approach to handling subgroups to ensure we can differentiate between subgroups amongst multiple series.

Anugraha-sutara committed:
Refactor web version,hide the filter button

Anugraha-sutara committed:
Commented out the filter  E2E test case since the UI element is no longer present.

Anugraha-sutara committed:
fix app crash

Anugraha-sutara committed:
fix app crash

mikedawson committed:
Add initial interfaces for graph value formatting.

mikedawson committed:
Add work in progress implementation of formatters.

poojaustad committed:
updated report description

mikedawson commented:
Behavior note: delete series button should only be visible if there is more than one series in graph

Add to uiState something like
```
 val deleteSeriesButtonVisible: Boolean
    get() = series.size > 1
```

Anugraha-sutara committed:
fix app crash and use formatter through di

Anugraha-sutara committed:
Merge remote-tracking branch 'origin/dev-customizable-reports-draw-graph-from-data-output' into dev-customizable-reports-draw-graph-from-data-output

Anugraha-sutara committed:
Add refactor

Anugraha-sutara committed:
add formatter on web side

Anugraha-sutara committed:
fix unit test

Anugraha-sutara committed:
remove commented code

Anugraha-sutara committed:
add refactor

Anugraha-sutara committed:
add series delete icon fix on web

mikedawson committed:
Add UiText sealed class.

Anugraha-sutara committed:
add string resource impl for graph
